tensorflow_quantum/core/qsim/state_space.cc

```diff
@@ -149,6 +150,58 @@ void StateSpace::SampleState(const int m, std::vector<uint64_t>* samples) {
   }
 }
 
+tensorflow::Status StateSpace::ComputeSampledExpectation(
+    const tfq::proto::PauliSum& p_sum, StateSpace* scratch,
+    float* expectation_value, const int m) {
+  // apply the  gates of the pauliterms to a copy of the wavefunction
+  // and add up expectation value term by term.
+  tensorflow::Status status = tensorflow::Status::OK();
+  for (const tfq::proto::PauliTerm& term : p_sum.terms()) {
+    // catch identity terms
+    if (term.paulis_size() == 0) {
+      *expectation_value += term.coefficient_real();
+      // TODO(zaqqwerty): error somewhere if identities have any imaginary part
+      continue;
+    }
+
+    // Transform state into the measurement basis and sample it
+    Circuit measurement_circuit;
+    status =
+        ZBasisCircuitFromPauliTerm(term, num_qubits_, &measurement_circuit);
+    if (!status.ok()) {
+      return status;
+    }
+    scratch->CopyFrom(*this);
+    status = scratch->Update(measurement_circuit);
+    if (!status.ok()) {
+      return status;
+    }
+    std::vector<uint64_t> state_samples;
+    scratch->SampleState(m, &state_samples);
+
+    // Find qubits on which to measure parity
+    std::vector<unsigned int> parity_bits;
+    for (const tfq::proto::PauliQubitPair& pair : term.paulis()) {
+      unsigned int location;
+      if (!absl::SimpleAtoi(pair.qubit_id(), &location)) {
+        return tensorflow::Status(
+            tensorflow::error::INVALID_ARGUMENT,
+            "Could not parse Pauli term qubit id: " + pair.ShortDebugString());
```

Comment:
Not entirely sure we need this since any PauliSum objects coming in most likely came from `ResolveQubitIds` which already puts things in the right format, but this isn't a huge deal I guess.

Reply:
I find without a check we get the warning `ABSL_MUST_USE_RESULT bool SimpleAtoi(absl::string_view s, int_type* out)`

Reply:
Fair enough, can't hurt to do an extra safety check here and there.